### PR TITLE
Convert selective checks to Breeze Python

### LINKED ISSUE #19971: CI: Rewrite selective check script in Python
Breeze's CI uses selective check script, to determine which parts of the tests and build should be executed. This helps to save a LOT of build time for simpler PRs - for example when there is a PR with just a documentation change, it wil not run unit tests.

It's used in two places: ci.yml and build.yml:

* https://github.com/apache/airflow/blob/eaa8ac72fc901de163b912a94dbe675045d2a009/.github/workflows/ci.yml#L209
* https://github.com/apache/airflow/blob/eaa8ac72fc901de163b912a94dbe675045d2a009/.github/workflows/build-images.yml#L124

The logic for that is described in https://github.com/apache/airflow/blob/main/PULL_REQUEST_WORKFLOW.rst#id1

This boils down to simply generating the right set of outputs, based on what files were modified in the PR and what labels the PR has.

This should be rewritten in python and we should be able to remove and replace the selective check script in the CI of airflwo when this is complete.

### PR DESCRIPTION
**Pull request by potiuk**
Instead of bash-based, complex logic script to perform PR selective
checks we now integrated the whole logic into Breeze Python code.

It is now much simplified, when it comes to algorithm. We've
implemented simple rule-based decision tree. The rules describing
the decision tree are now are now much easier
to reason about and they correspond one-to-one with the rules
that are implemented in the code in rather straightforward way.

The code is much simpler and diagnostics of the selective checks
has also been vastly improved:

* The rule engine displays status of applying each rule and
  explains (with yellow warning message what decision was made
  and why. Informative messages are printed showing the resulting
  output

* List of files impacting the decision are also displayed

* The names of "ci file group" and "test type" were aligned

* Unit tests covering wide range of cases are added. Each test
  describes what is the case they demonstrate

* `breeze selective-checks` command that is used in CI can also
  be used locally by just providing commit-ish reference of the
  commit to check. This way you can very easily debug problems and
  fix them

Fixes: #19971

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
This is the next step in Breeze conversion. This time -800 lines of Bash Code (replaced with Python but including lots of unit tests) which is far more readable and easy to reason about.

Plus very clear diagnosticts of what decisions are made and why, including `breeze static-check` tool that can be used locally to debug and reproduce selective check behaviour with any local commit.

Regular PR that triggers all tests (core change) but does not  run any Helm/K8S/UI javascript tests

image

More complex change (this one) that trigers all possible tests because environment changed:
image

Fragment of more complex change (amazon + cncf.kubernetes) that triggers only subset of unit tests ("Always", "Providers") bit also all our K8S changes:

image

Tests are pretty comprehensive and we will be able to easily test and add any new cases in the future (especially when we split providers and want to make more fine-grained decisions which tests to run.

**Comment by potiuk** [comment]
cc: @edithturn @bowrna : -800 lines of bash less and much easier "logic" to grasp for selective checks implementation

**Comment by edithturn** [comment]
I wouldn't have thought of all those changes, really impressive @potiuk.

**Comment by potiuk** [comment]
> I wouldn't have thought of all those changes, really impressive @potiuk.

Just Python is way easier to write things better :)

**Comment by potiuk** [comment]
BTW. By converting the checks to Python I also found that current selectiuve checks were not "selective enough" as I missed the cases where "Helm" tests are run unnecessarily in quite a number of cases (For example they we run when only providers were modified - this should speed up PRs from contributors who only modified one or more providers and did not touch the code).

File: SELECTIVE_CHECKS.md [hunk 1]
```diff
@@ -21,20 +21,38 @@
 
 In order to optimise our CI jobs, we've implemented optimisations to only run selected checks for some
 kind of changes. The logic implemented reflects the internal architecture of Airflow 2.0 packages
-and it helps to keep down both the usage of jobs in GitHub Actions as well as CI feedback time to
+and it helps to keep down both the usage of jobs in GitHub Actions abd CI feedback time to
```
Comment (mik-laj) [inline SELECTIVE_CHECKS.md]: ?

File: dev/breeze/src/airflow_breeze/commands/ci_commands.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        )
+        return ()
+    changed_files = tuple(result.stdout.splitlines()) if result.stdout else ()
+    get_console().print("\n[info]Changed files:[/]\n")
+    get_console().print(changed_files)
+    get_console().print()
+    return changed_files
+
+
+@main.command(name="selective-check", help="Check if available docker resources are enough.")
```
Comment (mik-laj) [inline dev/breeze/src/airflow_breeze/commands/ci_commands.py]: Can you update description?

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/commands/ci_commands.py]: Right! 🤦 . Fixed!

**Comment by potiuk** [comment]
All Green!

**Review by uranusjr — APPROVED** [review]

**Comment by edithturn** [comment]
It was merged, well done @potiuk Jarek 🥳

**Pull request merged**